File: Models/SpeechResource.cs
using System;
using System.Collections.Generic;
using System.Text.Json.Serialization;

namespace TrueFluentPro.Models
{
    [Flags]
    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum SpeechCapability
    {
        None = 0,
        RealtimeSpeechToText = 1,
        BatchSpeechToText = 2,
        TextToSpeech = 4
    }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum SpeechVendorType
    {
        Microsoft,
        OpenAI,
        Tencent,
        Alibaba,
        Other
    }

    [JsonConverter(typeof(JsonStringEnumConverter))]
    public enum SpeechConnectorType
    {
        MicrosoftSpeech,
        AiSpeech,
        CustomSpeech
    }

    public class SpeechResource
    {
        public string Id { get; set; } = Guid.NewGuid().ToString();
        public string Name { get; set; } = "";
        public SpeechVendorType Vendor { get; set; } = SpeechVendorType.Microsoft;
        public SpeechConnectorType ConnectorType { get; set; } = SpeechConnectorType.MicrosoftSpeech;
        public bool IsEnabled { get; set; } = true;
        public SpeechCapability Capabilities { get; set; } = SpeechCapability.None;

        // 兼容微软 Speech 资源的连接信息。后续若扩展更多传统 speech 厂商，可再抽成独立详情模型。
        public string SubscriptionName { get; set; } = "";
        public string SubscriptionKey { get; set; } = "";
        public string ServiceRegion { get; set; } = "";
        public string Endpoint { get; set; } = "";

        // AI 语音资源只引用模型，不重复保存终结点连接信息。
        public ModelReference? RealtimeSpeechToTextModelRef { get; set; }
        public ModelReference? BatchSpeechToTextModelRef { get; set; }
        public ModelReference? TextToSpeechModelRef { get; set; }

        public static SpeechResource CreateMicrosoftResource(AzureSubscription subscription, int index)
        {
            return new SpeechResource
            {
                Id = BuildLegacyMicrosoftResourceId(index),
                Name = string.IsNullOrWhiteSpace(subscription.Name) ? $"Microsoft Speech {index + 1}" : subscription.Name.Trim(),
                Vendor = SpeechVendorType.Microsoft,
                ConnectorType = SpeechConnectorType.MicrosoftSpeech,
                IsEnabled = true,
                Capabilities = SpeechCapability.RealtimeSpeechToText | SpeechCapability.BatchSpeechToText,
                SubscriptionName = subscription.Name?.Trim() ?? "",
                SubscriptionKey = subscription.SubscriptionKey?.Trim() ?? "",
                ServiceRegion = subscription.GetEffectiveRegion(),
                Endpoint = subscription.GetEffectiveEndpoint()
            };
        }

        public static SpeechResource? CreateLegacyAiResource(
            ModelReference? realtimeModelRef,
            ModelReference? batchModelRef,
            ModelReference? ttsModelRef)
        {
            var capabilities = SpeechCapability.None;
            if (realtimeModelRef != null)
            {
                capabilities |= SpeechCapability.RealtimeSpeechToText;
            }

            if (batchModelRef != null)
            {
                capabilities |= SpeechCapability.BatchSpeechToText;
            }

            if (ttsModelRef != null)
            {
                capabilities |= SpeechCapability.TextToSpeech;
            }

            if (capabilities == SpeechCapability.None)
            {
                return null;
            }

            return new SpeechResource
            {
                Id = LegacyAiResourceId,
                Name = "默认 AI 语音资源",
                Vendor = SpeechVendorType.OpenAI,
                ConnectorType = SpeechConnectorType.AiSpeech,
                IsEnabled = true,
                Capabilities = capabilities,
                RealtimeSpeechToTextModelRef = CloneReference(realtimeModelRef),
                BatchSpeechToTextModelRef = CloneReference(batchModelRef),
                TextToSpeechModelRef = CloneReference(ttsModelRef)
            };
        }

        public static string BuildLegacyMicrosoftResourceId(int index) => $"legacy-microsoft-{index}";

        public const string LegacyAiResourceId = "legacy-ai-speech-default";

        public bool HasCapability(SpeechCapability capability)
            => (Capabilities & capability) == capability;

        public string GetDisplayName()
        {
            if (ConnectorType == SpeechConnectorType.MicrosoftSpeech)
            {
                var region = !string.IsNullOrWhiteSpace(ServiceRegion)
                    ? ServiceRegion
                    : AzureSubscription.ParseRegionFromEndpoint(Endpoint) ?? string.Empty;

                if (!string.IsNullOrWhiteSpace(region))
                {
                    return $"{Name} ({region})";
                }
            }

            return Name;
        }

        public bool TryCreateAzureSubscription(out AzureSubscription? subscription)
        {
            subscription = null;
            if (ConnectorType != SpeechConnectorType.MicrosoftSpeech)
            {
                return false;
            }

            subscription = new AzureSubscription
            {
                Name = string.IsNullOrWhiteSpace(SubscriptionName) ? Name : SubscriptionName,
                SubscriptionKey = SubscriptionKey,
                ServiceRegion = ServiceRegion,
                Endpoint = Endpoint
            };

            return true;
        }

        public SpeechResource Clone()
        {
            return new SpeechResource
            {
                Id = Id,
                Name = Name,
                Vendor = Vendor,
                ConnectorType = ConnectorType,
                IsEnabled = IsEnabled,
                Capabilities = Capabilities,
                SubscriptionName = SubscriptionName,
                SubscriptionKey = SubscriptionKey,
                ServiceRegion = ServiceRegion,
                Endpoint = Endpoint,
                RealtimeSpeechToTextModelRef = CloneReference(RealtimeSpeechToTextModelRef),
                BatchSpeechToTextModelRef = CloneReference(BatchSpeechToTextModelRef),
                TextToSpeechModelRef = CloneReference(TextToSpeechModelRef)
            };
        }

        public IEnumerable<string> GetReferencedEndpointIds()
        {
            var ids = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
            AddReferenceId(ids, RealtimeSpeechToTextModelRef);
            AddReferenceId(ids, BatchSpeechToTextModelRef);
            AddReferenceId(ids, TextToSpeechModelRef);
            return ids;
        }

        private static void AddReferenceId(HashSet<string> ids, ModelReference? reference)
        {
            if (!string.IsNullOrWhiteSpace(reference?.EndpointId))
            {
                ids.Add(reference.EndpointId);
            }
        }

        private static ModelReference? CloneReference(ModelReference? reference)
        {
            if (reference == null)
            {
                return null;
            }

            return new ModelReference
            {
                EndpointId = reference.EndpointId,
                ModelId = reference.ModelId
            };
        }
    }
}

File: Rust/rustnew-backend/crates/core/src/speech_resource.rs
//! 语音资源模型。
//!
//! 对应 C# `Models/SpeechResource.cs`，描述一个语音服务连接（厂商、连接类型、能力、凭据）。
//! 实时翻译先只用到 Microsoft Speech；其余厂商/连接类型保留以便后续扩展。

use serde::{Deserialize, Serialize};

use crate::capability::SpeechCapability;

/// 语音厂商类型。序列化值与 C# 枚举名保持一致。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
pub enum SpeechVendorType {
    #[default]
    Microsoft,
    #[serde(rename = "OpenAI")]
    OpenAi,
    Tencent,
    Alibaba,
    Other,
}

/// 语音连接类型。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
pub enum SpeechConnectorType {
    /// 原生 Microsoft 语音 SDK（密钥 + 区域 / 终结点）
    #[default]
    MicrosoftSpeech,
    /// 走 AI 网关 / OpenAI 兼容协议的语音
    AiSpeech,
    /// 自定义
    CustomSpeech,
}

/// 鉴权方式。
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SpeechAuthMode {
    #[default]
    ApiKey,
    Aad,
}

/// 一个语音资源。
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpeechResource {
    #[serde(default)]
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub vendor: SpeechVendorType,
    #[serde(default)]
    pub connector_type: SpeechConnectorType,
    #[serde(default = "default_true")]
    pub is_enabled: bool,
    #[serde(default)]
    pub capabilities: SpeechCapability,
    #[serde(default)]
    pub auth_mode: SpeechAuthMode,
    #[serde(default)]
    pub subscription_name: String,
    #[serde(default)]
    pub subscription_key: String,
    #[serde(default)]
    pub service_region: String,
    #[serde(default)]
    pub endpoint: String,
}

fn default_true() -> bool {
    true
}

impl Default for SpeechResource {
    fn default() -> Self {
        Self {
            id: new_resource_id(),
            name: String::new(),
            vendor: SpeechVendorType::Microsoft,
            connector_type: SpeechConnectorType::MicrosoftSpeech,
            is_enabled: true,
            capabilities: SpeechCapability::REALTIME_SPEECH_TO_TEXT,
            auth_mode: SpeechAuthMode::ApiKey,
            subscription_name: String::new(),
            subscription_key: String::new(),
            service_region: String::new(),
            endpoint: String::new(),
        }
    }
}

impl SpeechResource {
    /// 创建一个 Microsoft 实时语音资源。
    pub fn new_microsoft(name: impl Into<String>, key: impl Into<String>, region: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            subscription_key: key.into(),
            service_region: region.into(),
            capabilities: SpeechCapability::REALTIME_SPEECH_TO_TEXT
                | SpeechCapability::BATCH_SPEECH_TO_TEXT,
            ..Default::default()
        }
    }

    /// 凭据是否完整可用。
    pub fn is_valid(&self) -> bool {
        if self.subscription_key.trim().is_empty() {
            return false;
        }
        !self.service_region.trim().is_empty() || !self.endpoint.trim().is_empty()
    }

    /// 是否中国区终结点（`.azure.cn`）。
    pub fn is_china(&self) -> bool {
        let ep = self.endpoint.to_ascii_lowercase();
        ep.contains(".azure.cn") || self.service_region.to_ascii_lowercase().starts_with("china")
    }

    /// 有效区域：优先 service_region，否则尝试从终结点解析。
    pub fn effective_region(&self) -> Option<String> {
        let r = self.service_region.trim();
        if !r.is_empty() {
            return Some(r.to_string());
        }
        parse_region_from_endpoint(&self.endpoint)
    }
}

/// 从形如 `wss://southeastasia.stt.speech.microsoft.com` /
/// `https://southeastasia.api.cognitive.microsoft.com` 的终结点中解析区域。
pub fn parse_region_from_endpoint(endpoint: &str) -> Option<String> {
    let ep = endpoint.trim();
    if ep.is_empty() {
        return None;
    }
    let without_scheme = ep
        .split_once("://")
        .map(|(_, rest)| rest)
        .unwrap_or(ep);
    let host = without_scheme.split(['/', ':']).next().unwrap_or("");
    let first = host.split('.').next().unwrap_or("");
    if first.is_empty() {
        None
    } else {
        Some(first.to_string())
    }
}

/// 生成一个时间可排序的资源 ID（ULID 风格，Crockford Base32）。
pub fn new_resource_id() -> String {
    crate::storage::new_ulid()
}
